File: scripts/purge-openai-stored-responses.py
#!/usr/bin/env python3
"""
Purge stored OpenAI /v1/responses records (ai-therapist-188).

WHY THIS EXISTS
---------------
The Responses API defaults `store` to true. Our redaction passes never set it
false, so every redaction call — the one payload in the system that is by
definition RAW, PRE-REDACTION participant transcript — was retained as
application state on OpenAI's side. The leak was fixed in commit f4c0234 and
verified stopped; this script removes what had already accumulated.

WHY IT NEEDS ITS OWN KEY
------------------------
Listing stored responses rejects an ordinary project key outright:
    "must be made with a session key ... it can only be made from the browser"
But an ADMIN key gets a different, far more useful error:
    "Missing scopes: api.responses.read ... if you're using a restricted API
     key, that it has the necessary scopes."
So a *restricted* admin key carrying `api.responses.read` (list) and
`api.responses.write` (delete) can do the whole job over plain HTTPS — no
browser scripting, no session-token handling.

CREATE THE KEY
--------------
  platform.openai.com/settings/organization/admin-keys -> Create new admin key
    Permissions: Restricted
    Enable:      api.responses.read   AND   api.responses.write
  This key is powerful and single-purpose: DELETE IT when the purge is done.
  Do NOT reuse the app's OPENAI_ADMIN_KEY — that one is read-only billing on
  purpose and should never be able to mutate anything.

USAGE
-----
  python3 scripts/purge-openai-stored-responses.py --key-file /tmp/purge-key
  python3 scripts/purge-openai-stored-responses.py            # hidden prompt

  --dry-run     list and sample only, delete nothing  (DEFAULT)
  --execute     actually delete
  --workers N   parallel deletes (default 8)

Re-running is safe: ids already gone return 404 and count as done.
"""
import argparse
import concurrent.futures as futures
import getpass
import json
import sys
import threading
import time
import urllib.error
import urllib.request

API = "https://api.openai.com/v1/responses"


def call(method, url, key, timeout=60):
    """Return (status, parsed_body_or_text). Never raises on HTTP errors."""
    req = urllib.request.Request(url, method=method)
    req.add_header("Authorization", f"Bearer {key}")
    try:
        with urllib.request.urlopen(req, timeout=timeout) as r:
            raw = r.read().decode("utf-8", "replace")
            try:
                return r.status, json.loads(raw)
            except json.JSONDecodeError:
                return r.status, raw
    except urllib.error.HTTPError as e:
        raw = e.read().decode("utf-8", "replace")
        try:
            return e.code, json.loads(raw)
        except json.JSONDecodeError:
            return e.code, raw
    except Exception as e:                     # network-level
        return 0, str(e)


def preflight(key):
    """Fail loudly and specifically rather than half-running."""
    status, body = call("GET", f"{API}?limit=1", key)
    if status == 200:
        return
    msg = ""
    if isinstance(body, dict):
        msg = body.get("error", {}).get("message", "")
    if "session key" in msg:
        sys.exit(
            "ERROR: that looks like a PROJECT key. Listing needs an ADMIN key with\n"
            "       the api.responses.read scope. See the header of this file."
        )
    if "Missing scopes" in msg:
        sys.exit(
            f"ERROR: key is missing a required scope.\n       {msg}\n"
            "       Recreate it as Restricted with api.responses.read AND\n"
            "       api.responses.write."
        )
    sys.exit(f"ERROR: preflight failed (HTTP {status}): {str(msg or body)[:300]}")


def collect(key):
    """Page through every stored response id."""
    ids, sample, after = [], [], None
    while True:
        url = f"{API}?limit=100" + (f"&after={after}" if after else "")
        status, body = call("GET", url, key)
        if status == 429:
            time.sleep(3)
            continue
        if status != 200:
            print(f"  ! list stopped at {len(ids)} ids (HTTP {status}): "
                  f"{str(body)[:200]}", file=sys.stderr)
            break
        rows = body.get("data") or []
        if not rows:
            break
        for row in rows:
            ids.append(row["id"])
            if len(sample) < 6:
                created = row.get("created_at")
                sample.append({
                    "id": row["id"],
                    "model": row.get("model"),
                    "created": time.strftime("%Y-%m-%d %H:%M",
                                             time.localtime(created)) if created else "?",
                })
        after = rows[-1]["id"]
        print(f"  ...{len(ids)} collected", flush=True)
        if not body.get("has_more"):
            break
    return ids, sample


def purge(ids, key, workers):
    done = failed = 0
    lock = threading.Lock()
    total = len(ids)

    def one(rid):
        nonlocal done, failed
        for _ in range(6):
            status, _body = call("DELETE", f"{API}/{rid}", key)
            if status == 429:
                time.sleep(2)
                continue
            # 404 == already gone; count it so re-runs converge cleanly.
            with lock:
                if status in (200, 204, 404):
                    done += 1
                else:
                    failed += 1
                n = done + failed
                if n % 250 == 0 or n == total:
                    print(f"  {n}/{total}  deleted={done} failed={failed}", flush=True)
            return
        with lock:
            failed += 1

    with futures.ThreadPoolExecutor(max_workers=workers) as pool:
        list(pool.map(one, ids))
    return done, failed


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--key-file", help="file containing the scoped admin key")
    ap.add_argument("--execute", action="store_true", help="actually delete")
    ap.add_argument("--dry-run", action="store_true", default=None)
    ap.add_argument("--workers", type=int, default=8)
    args = ap.parse_args()

    if args.key_file:
        key = open(args.key_file).read().strip()
    else:
        key = getpass.getpass("Paste the scoped admin key (hidden): ").strip()
    if not key.startswith("sk-admin-"):
        sys.exit("ERROR: expected an admin key (sk-admin-...).")

    print("==> preflight")
    preflight(key)
    print("    OK - key can list stored responses.")

    print("==> collecting ids")
    ids, sample = collect(key)
    print(f"\nFound {len(ids)} stored responses.")
    for s in sample:
        print(f"    {s['id']}  {s['model']}  {s['created']}")
    if not ids:
        print("Nothing to delete.")
        return

    if not args.execute:
        print("\nDRY RUN - nothing deleted. Re-run with --execute to purge.")
        return

    print(f"\n==> deleting {len(ids)} records with {args.workers} workers")
    done, failed = purge(ids, key, args.workers)
    print(f"\nDeleted {done}, failed {failed}.")

    # Verify rather than trust.
    print("==> verifying")
    remaining, _ = collect(key)
    print(f"Remaining stored responses: {len(remaining)}")
    if remaining:
        print("Re-run to clear the stragglers (deletes are idempotent).")
        sys.exit(1)
    print("All stored responses purged.")


if __name__ == "__main__":
    main()

File: scripts/purge-openai-stored-responses.js
/* ============================================================================
 * PURGE STORED OPENAI RESPONSES  (ai-therapist-188)  — browser console snippet
 * ============================================================================
 * Deletes the /v1/responses records that accumulated while the Responses API
 * `store` param defaulted to true — including every redaction pass, which
 * carried RAW PRE-REDACTION participant transcripts. The leak itself was fixed
 * in commit f4c0234 and verified stopped; this clears the backlog.
 *
 * WHY THIS MUST RUN IN THE BROWSER
 * --------------------------------
 * Verified exhaustively on 2026-09-09 — there is no server-side path:
 *   - project key (sk-proj-), every header/param variant:
 *       "must be made with a session key (that is, it can only be made from
 *        the browser)"
 *   - admin key (sk-admin-) with UNRESTRICTED permissions, every variant:
 *       "Missing scopes: api.responses.read"
 *     Unrestricted already grants every admin scope, so a Restricted key
 *     cannot have more: api.responses.* is not an admin scope at all.
 *   - the console Logs UI has no delete affordance (no trash, no bulk select)
 *   - we never persisted the resp_... ids, so DELETE (which a project key CAN
 *     call) has nothing to iterate over
 * Listing is session-only by design. Hence: run it in the logged-in tab.
 *
 * HOW TO RUN
 * ----------
 *  1. Open  https://platform.openai.com/logs   (logged in)
 *  2. DevTools console:  Cmd+Option+J
 *  3. Paste this entire file, press Enter
 *  4. Keep the tab OPEN and FOCUSED until it prints DONE — background tabs get
 *     throttled and this will crawl.
 *
 * It runs a DRY RUN first and deletes nothing. Read the sample it prints, then
 * set DRY_RUN = false at the top and paste again to actually purge.
 *
 * If you see a 401: the console is using a bearer token rather than cookies.
 *   DevTools > Network > click the `responses?include[]=...` request >
 *   Headers > copy the `authorization:` value > set AUTH_OVERRIDE below.
 *   That token is a live credential — keep it out of chats, tickets, commits.
 *
 * Deletion is irreversible and these records exist nowhere else. Re-running is
 * safe: already-deleted ids return 404 and count as done, so it converges.
 * ========================================================================= */

(async () => {
  const DRY_RUN       = true;   // <-- set false to actually delete
  const AUTH_OVERRIDE = null;   // <-- 'Bearer sess-...' only if you get a 401
  const CONCURRENCY   = 6;
  const PAGE_SIZE     = 100;

  const API  = 'https://api.openai.com/v1/responses';
  const opts = {
    credentials: 'include',
    headers: AUTH_OVERRIDE ? { authorization: AUTH_OVERRIDE } : {},
  };
  const sleep = ms => new Promise(r => setTimeout(r, ms));

  // ---- preflight -----------------------------------------------------------
  const probe = await fetch(`${API}?limit=1`, opts);
  if (probe.status === 401 || probe.status === 403) {
    console.error('%cAuth failed (' + probe.status + ')',
      'color:red;font-weight:bold',
      '\nSet AUTH_OVERRIDE — see the header comment.');
    return;
  }
  if (!probe.ok) {
    console.error('Unexpected status', probe.status,
      (await probe.text()).slice(0, 300));
    return;
  }

  // ---- enumerate -----------------------------------------------------------
  console.log('%cCollecting response ids…', 'font-weight:bold');
  const ids = [], sample = [];
  let after = null;
  for (;;) {
    const res = await fetch(
      `${API}?limit=${PAGE_SIZE}${after ? `&after=${after}` : ''}`, opts);
    if (res.status === 429) { await sleep(3000); continue; }
    if (!res.ok) {
      console.error('List stopped at', ids.length, 'ids —', res.status,
        (await res.text()).slice(0, 200));
      break;
    }
    const page = await res.json();
    const rows = page.data || [];
    if (!rows.length) break;
    for (const row of rows) {
      ids.push(row.id);
      if (sample.length < 6) sample.push({
        id: row.id,
        model: row.model,
        created: row.created_at
          ? new Date(row.created_at * 1000).toLocaleString() : '?',
      });
    }
    after = rows[rows.length - 1].id;
    console.log(`  …${ids.length} collected`);
    if (!page.has_more) break;
  }

  console.log(`%cFound ${ids.length} stored responses.`, 'font-weight:bold');
  console.table(sample);
  if (!ids.length) { console.log('Nothing to do.'); return; }

  if (DRY_RUN) {
    console.log('%cDRY RUN — nothing deleted.', 'color:orange;font-weight:bold',
      '\nConfirm the sample above is your data, then set DRY_RUN = false and re-paste.');
    return;
  }

  // ---- delete --------------------------------------------------------------
  console.log('%cDeleting…', 'color:red;font-weight:bold');
  let done = 0, failed = 0;
  const queue = ids.slice();
  const worker = async () => {
    while (queue.length) {
      const id = queue.pop();
      try {
        const res = await fetch(`${API}/${id}`, { method: 'DELETE', ...opts });
        if (res.status === 429) { queue.push(id); await sleep(2000); continue; }
        if (res.ok || res.status === 404) done++; else failed++;  // 404 = already gone
      } catch { failed++; }
      const n = done + failed;
      if (n % 250 === 0) console.log(`  ${n}/${ids.length}  deleted=${done} failed=${failed}`);
    }
  };
  await Promise.all(Array.from({ length: CONCURRENCY }, worker));

  // ---- verify rather than trust the counter --------------------------------
  const check = await fetch(`${API}?limit=1`, opts);
  const left  = check.ok ? ((await check.json()).data || []).length : '?';
  console.log(`%cDONE — deleted ${done}, failed ${failed}. Remaining on first page: ${left}`,
    'color:green;font-weight:bold');
  if (failed || left) console.log('Re-paste to sweep stragglers (deletes are idempotent).');
})();
